multi-int-store/src/main/java/com/connexta/multiintstore/controllers/RetrieveController.java

```diff
@@ -6,32 +6,48 @@
  */
 package com.connexta.multiintstore.controllers;
 
+import com.connexta.multiintstore.adaptors.RetrieveResponse;
+import com.connexta.multiintstore.common.ProductStorageManager;
+import com.connexta.multiintstore.services.api.RetrieveApi;
+import javax.validation.constraints.NotNull;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.core.io.Resource;
+import org.springframework.http.HttpHeaders;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.ResponseEntity;
-import org.springframework.web.bind.annotation.GetMapping;
-import org.springframework.web.bind.annotation.PathVariable;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
 
+@Slf4j
 @RestController
 @RequestMapping("/mis")
-public class RetrieveController {
-  @GetMapping(value = "/product/{productId}")
-  public ResponseEntity retrieveProduct(@PathVariable("productId") String productId) {
-    return ResponseEntity.status(HttpStatus.NOT_IMPLEMENTED)
-        .body("Yay, you retrieved a product with ID: " + productId + "!!");
+public class RetrieveController implements RetrieveApi {
+
+  @NotNull private final ProductStorageManager productStorageManager;
+
+  public RetrieveController(@NotNull ProductStorageManager productStorageManager) {
+    this.productStorageManager = productStorageManager;
   }
 
-  @GetMapping(value = "/product/{productId}/{metadataType}")
-  public ResponseEntity retrieveMetadata(
-      @PathVariable("productId") String productId,
-      @PathVariable("metadataType") String metadataType) {
-    return ResponseEntity.status(HttpStatus.NOT_IMPLEMENTED)
-        .body(
-            "Yay, you retrieved "
-                + metadataType
-                + " for the the product with ID: "
-                + productId
-                + "!!");
+  @Override
+  public ResponseEntity<Resource> retrieveProduct(String productId) {
+    final RetrieveResponse retrieveResponse;
+
+    try {
+      retrieveResponse = productStorageManager.retrieveProduct(productId);
+    } catch (RuntimeException e) {
+      log.warn("Unable to retrieve {}", productId, e);
+      return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+
+    log.info("Successfully retrieved id={}", productId);
+
+    final Resource resource = retrieveResponse.getResource();
+    return ResponseEntity.ok()
+        .contentType(retrieveResponse.getMediaType())
+        .header(
+            HttpHeaders.CONTENT_DISPOSITION,
```

Comment:
Use Spring's [ContentDisposition](https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/http/ContentDisposition.html) class to build the header.